pkg/controller/crdmirroring/crdhandler/clustergroup.go

```diff
@@ -0,0 +1,163 @@
+	crdclientset "github.com/vmware-tanzu/antrea/pkg/client/clientset/versioned"
+	crdlister "github.com/vmware-tanzu/antrea/pkg/client/listers/crd/v1alpha2"
+	"github.com/vmware-tanzu/antrea/pkg/controller/crdmirroring/types"
+	legacycore "github.com/vmware-tanzu/antrea/pkg/legacyapis/core/v1alpha2"
+	legacycrdclientset "github.com/vmware-tanzu/antrea/pkg/legacyclient/clientset/versioned"
+	legacycorelister "github.com/vmware-tanzu/antrea/pkg/legacyclient/listers/core/v1alpha2"
+)
+
+type ClusterGroupHandler struct {
+	lister       crdlister.ClusterGroupLister
+	legacyLister legacycorelister.ClusterGroupLister
+	client       crdclientset.Interface
+	legacyClient legacycrdclientset.Interface
+}
+
+func NewClusterGroupHandler(lister crdlister.ClusterGroupLister,
+	legacyLister legacycorelister.ClusterGroupLister,
+	client crdclientset.Interface,
+	legacyClient legacycrdclientset.Interface) types.MirroringHandler {
+	mc := &ClusterGroupHandler{
+		lister:       lister,
+		legacyLister: legacyLister,
+		client:       client,
+		legacyClient: legacyClient,
+	}
+	return mc
+}
+
+// GetNewObject gets the mirroring new ClusterGroup struct
+func (c *ClusterGroupHandler) GetNewObject(namespace, name string) (metav1.Object, error) {
+	lister := c.lister
+	return lister.Get(name)
+}
+
+// AddNewObject creates the mirroring new ClusterGroup
+func (c *ClusterGroupHandler) AddNewObject(obj metav1.Object) error {
+	l := obj.(*legacycore.ClusterGroup)
+	n := c.buildNewObject(l)
+	client := c.client.CrdV1alpha2().ClusterGroups()
+	_, err := client.Create(context.TODO(), n, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// SyncObject updates the mirroring new ClusterGroup
+func (c *ClusterGroupHandler) SyncObject(legacyObj, newObj metav1.Object) error {
+	if !c.deepEqualSpecAndLabels(legacyObj, newObj) {
+		n := c.syncNewObject(legacyObj, newObj)
+		newClient := c.client.CrdV1alpha2().ClusterGroups()
+		_, err := newClient.Update(context.TODO(), n, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+	}
+
+	if !c.deepEqualStatus(legacyObj, newObj) {
+		l := c.syncLegacyObject(legacyObj, newObj)
+		legacyClient := c.legacyClient.CoreV1alpha2().ClusterGroups()
+		_, err := legacyClient.UpdateStatus(context.TODO(), l, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+//DeleteNewObject deletes the mirroring new ClusterGroup
+func (c *ClusterGroupHandler) DeleteNewObject(namespace, name string) error {
+	client := c.client.CrdV1alpha2().ClusterGroups()
+	return client.Delete(context.TODO(), name, metav1.DeleteOptions{})
+}
+
+// StaleLegacyObject updates the mirroring new ClusterGroup
+func (c *ClusterGroupHandler) StaleLegacyObject(legacyObj, newObj metav1.Object) error {
+	delete(newObj.GetAnnotations(), types.ManagedBy)
+	newClient := c.client.CrdV1alpha2().ClusterGroups()
+	n := newObj.(*crd.ClusterGroup)
+	_, err := newClient.Update(context.TODO(), n, metav1.UpdateOptions{})
```

Comment:
I think most code in this method can be moved to controller as I suggested before. There is only one difference, the one line code to call the new client API.

Reply:
how about this comment? Also, staleLegacyObject doesn't sound proper. By extracting the logic to controller, we just need a UpdateNewObject method.

Reply:
Sorry, I forgot this comment. Using UpdateNewObject is better.

Reply:
still don't see this is updated except the method name is updated. This could have been one line method if the DeepCopy and annotation mutation is moved to controller

Reply:
All Objects implement runtime.Object which has DeepCopy method.